src/main/scala/utility/Sort.scala

```diff
@@ -0,0 +1,151 @@
+***************************************************************************************/
+
+package utility
+
+import chisel3._
+import chisel3.util._
+
+class TwithPtr[A <: Data, B <: CircularQueuePtr[B]](a: A, b: B) extends Bundle {
+  val valid = Bool()
+  val bits = a.cloneType
+  val ptr = b.cloneType
+}
+
+object TwithPtr {
+  /** Factory of class [[TwithPtr]]
+    *
+    * @param  valid Bool
+    * @param  bits the origin Data
+    * @param  ptr [[CircularQueuePtr]]
+    * @return new hardware of [[TwithPtr]]
+    */
+  def apply[A <: Data, B <: CircularQueuePtr[B]](valid: Bool, bits: A, ptr: B): TwithPtr[A, B] = {
+    val x = Wire(new TwithPtr[A, B](bits, ptr))
+    x.valid := valid
+    x.bits := bits
+    x.ptr := ptr
+    x
+  }
+}
+
+object HwSort {
+  /** Hardware Sort, a small combinational sorter that can sort up to 4 data by Ptr.
+    * 
+    * The sort is stable. That is, elements that are invalid or equal (as determined by `cmp` function),
+    * appear in the same order in the sorted sequence as in the original.
+    * 
+    * @tparam A default Data
+    * @tparam B the type where "<" is defined, here it refers to [[CircularQueuePtr]]
+    * @param  xVec the origin Vec data.
+    * @param  cmp the ordering assumed on domain `B`, ascending sorting is default.
+    * @return a Vec consisting of the elements of `xVec` sorted by `cmp`
+    * 
+    * @example {{{
+    *   // Ascending sorting is default, i.e. the oldest comes first.
+    *   val ascSorted = HwSort(VecInit(io.req.map { case x => TwithPtr(x.valid, x.bits, x.bits.uop.robIdx) }))
+    *   
+    *   // Descending sorting can be customized, i.e. the new one comes first.
+    *   // method1: implicitly set
+    *   implicit val customCmp: (RobPtr, RobPtr) => Bool = (x, y) => x > y
+    *   val desSorted1 = HwSort(VecInit(io.req.map { case x => TwithPtr(x.valid, x.bits, x.bits.uop.robIdx) }))
+    *   // method2: explicitly provide
+    *   val desSorted2 = HwSort(VecInit(io.req.map { case x => TwithPtr(x.valid, x.bits, x.bits.uop.robIdx) }))(
+    *     (a, b) => a > b
+    *   )
+    * }}}
+    */
+  def apply[A <: Data, B <: CircularQueuePtr[B]](xVec: Vec[TwithPtr[A, B]])
+  (implicit cmp: (B, B) => Bool = (x: B, y: B) => x < y)
+  : Vec[TwithPtr[A, B]] = {
+    var size = xVec.length
+    val res = WireInit(xVec)
+
+    if (size == 1) {
+      res := xVec
+    } else if (size == 2) {
+      // total ~20 ps
+      // 1 is older than 0 (only when 0 and 1 are both valid)
+      val swap = xVec(0).valid && xVec(1).valid && cmp(xVec(1).ptr, xVec(0).ptr)
+      when(swap) {
+        res(0) := xVec(1)
+        res(1) := xVec(0)
+      }
+    } else if (size == 3) {
+      // total ~40 ps
+      val row0 = WireInit(xVec)
+      val row1 = Wire(xVec.cloneType)
+      val row2 = Wire(xVec.cloneType)
+      val row3 = Wire(xVec.cloneType)
+
+      val tmp0 = apply(VecInit(row0.slice(0, 2)))
```

Comment:
Maybe better naming?

Reply:
such as? In my opinion, `row*` and `tmp*` are self-explanatory.